### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] false positive on type narrowing

### ISSUE THREAD

**Issue by cainlevy** [typescript-eslint/typescript-eslint#11242]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.3&fileType=.tsx&code=C4TwDgpgBAYgdlAvFA3gWAFBSgHgEpQQAewEcAJgM5SXABOAlnAOZQA%2BUArhRAGZMRyAPgAUASgBcUPAG5MAX0wBjAPZxaUYFPhIo4pEK49%2BcQXIyZe3JcAZqoAGxWUIAeToAZZxHFTajFnYoOE4AWwAjCDpUTGwGXj0AWQBDYAALADo6ZIoVUP0hZAAGDIBWMRisbCg6CGBOOgRgfWTqfyZmc2x5QgcXSuqauoam8S6oRQxJoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false

### Repro Code

```TypeScript
type Fn = {
  <R extends string | undefined>(): R;
}
const t: Fn = () => undefined;

function loseOrLose(): string | number {
  if (Math.random() >= 0.5) {
    // no-unnecessary-type-assertion thinks the "as string" is unnecessary
    return t() as string;
  } else {
    // but without `as string` typescript fails
    return t();
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true,
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected that `as string` would be allowed, since TypeScript fails without it.

### Actual Result

no-unnecessary-type-assertion claims the `as string` is unnecessary

### Additional Info

This is similar to https://github.com/typescript-eslint/typescript-eslint/issues/11054, but I do not know if it has the same root cause.

**Comment by bradzacher** [comment]
This is working as expected due to limitations in TS.

You have a generic that is only used in a return location. This causes TS to "backfill" the generic using the `as`'s type. [If you use intellisense you can see what I mean](https://typescript-eslint.io/play/#ts=5.8.3&fileType=.tsx&code=C4TwDgpgBAYgdlAvFA3gWAFBSgHgEpQQAewEcAJgM5SXABOAlnAOZQA%2BUArhRAGZMRyAPgAUASgBcUPAG5MAX0wBjAPZxaUYFPhIo4pEK49%2BcQXIyZe3JcAZqoAGxWUIAeToAZZxHFTajFnYoOE4AWwAjCDpUTGwGXj0AWQBDYAALADo6ZIoVUP0hZAAGDIBWMRisbCg6CGBOOgRgfWTqfyZmc2qoAHoe7qgAPQB%2BWKh5QgcXSu7a%2BsbNcS7xhUwgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false).

Because of this TS behaviour - when we inspect the types we see the signature of the function as `() => string`. So we error because based on the types you're doing `string as string`.

Of course if you remove the `as` then the function falls back to `() => string | undefined` - but we can't know that.

The fix is

a) don't use generics that are only used in return type locations. These generics are unsafe to begin with as they are impossible to implement correctly.
b) use the generic syntax instead `t<string>()`

**Comment by cainlevy** [comment]
Thanks for the reply. It sounds like there's nothing ESLint can do about this one. The logic of `no-unnecessary-type-assertion` seems to have a number of false positive cases due to limitations in TS.

It also appears there is no effective workaround:

a) The function I'm using is from i18next, and is unfortunately not something I can refactor. I trust they have reasons for the result type as it exists and that the function implementation is more interesting than the simplistic return statement in my reproduction.
b) When I use the generic syntax `t<string>()` I get a `no-unnecessary-type-arguments` violation because `string` is the default value for the generic parameter.
